### PR TITLE
changes DataLoaders.from_dsets

### PR DESCRIPTION
**Pull request by arampacha**
Linked issues #3396 #3373 
@riven314 and myself came up with possible solution. This PR changes `DataLoaders.from_dsets` to resamble `Datasets.dataloaders` ([link](https://github.com/fastai/fastai/blob/301016c5d3de2bdb5269121bd0716538d85f7409/fastai/data/core.py#L221)). This resolves aforementioned issue and allows to pass custom arguments to valid dataloader (e.g. val_res for `SortedDL`).

Open questions:
- How do we deal with situation when only one dataset is passed to constructor. As of now it will trigger error when trying to access `dls.valid`?
- It's still not quite clear what exactly was causing the issue. It was not observed with `do_setup=False` so probably something went wrong when setting up `Pipeline`s. But `Pipeline.setup` in isolation doesn't seem to cause same effect.

### CODE REVIEW

File: nbs/03_data.core.ipynb [hunk 1]
```diff
@@ -674,6 +679,29 @@
...
+     "execution_count": null,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "dls3 = DataLoaders.from_dsets(start, start, bs=4, val_bs=8, after_item=[_TestTfm()], val_after_batch=[_TestTfm()])\n",
+    "test_eq(len(dls3.train.after_batch.fs),0)\n",
+    "test_eq(len(dls3.valid.after_batch.fs),1)\n",
+    "dls.train.after_item, dls.valid.after_item"
```
Comment (riven314) [inline nbs/03_data.core.ipynb:702]: personally I think its better for this line to be a test:  
```
test_eq(len(dls.train.after_item.fs), len(dls.valid.after_item.fs))
```

Additionally, shall we also test `bs` is set properly on train & valid?  
```
test_eq(dls.train.bs, 4)
test_eq(dls.valid.bs, 8)
```

**Comment by jph00** [comment]
Why not just use `deepcopy`? A smaller change like that would be easier for me to review. Are there significant downsides?

**Comment by arampacha** [comment]
Using `deepcopy` was the first solution that came to both @riven314 and me and it fixes the issue :)

The solution proposed in this PR has advantage of handling kwargs specific to `valid` dataloader. Example use case when this can be useful - providing `val_res` for speeding up `SortedDL` creation. It is possible to modify current version of `.from_dsets` to take `val_*` keyword arguments as well. But that will also introduce substantial changes to review.
The proposed version makes `Dataloaders.from_dsets` more consistent with `Datasets.dataloaders`. These two have similar purpose and structure. In my opinion this should make those easier to maintain.

There is also an option to construct `Datasets` object from datasets passed to `Dataloaders.from_dsets` and use its `.dataloaders` method. But it might be unnecessary complication for the task. I can give it a try if you think it is a worthy idea.

**Comment by tcapelle** [comment]
why not just do something like this:
```
    @classmethod
    def from_dsets(cls, *ds, path='.',  bs=64, device=None, dl_type=TfmdDL, **kwargs):
        default = (True,) + (False,) * (len(ds)-1)
        defaults = {'shuffle': default, 'drop_last': default}
        tfms = {k:tuple(Pipeline(kwargs[k]) for i in range_of(ds)) for k in _batch_tfms if k in kwargs}
        kwargs = merge(defaults, {k: tuplify(v, match=ds) for k,v in kwargs.items() if k not in _batch_tfms}, tfms)
        kwargs = [{k: v[i] for k,v in kwargs.items()} for i in range_of(ds)]
        return cls(*[dl_type(d, bs=bs, **k) for d,k in zip(ds, kwargs)], path=path, device=device)
```

**Comment by arampacha** [comment]
@tcapelle, from previous message:
> The proposed version makes Dataloaders.from_dsets more consistent with Datasets.dataloaders. These two have similar purpose and structure. In my opinion this should make those easier to maintain.

I think this can help to avoid the kind of issues when we have an unexpected behavior in one method, while other method with similar functionality works fine. But that's just my opinion and it might be wrong. I can close the PR if proposed changes are not considered useful.

**Comment by tcapelle** [comment]
Maybe `from_dsets` should not exist, and the user should be pushed towards creating explicitly their dataloaders with transforms using the `TfmDL` constructor. If you have plain datasets without transforms on the input, the needed `kwargs` on `from_dsets` can quickly become messy if you want to customize the full pipeline.

**Comment by tcapelle** [comment]
The `from_df` method need reworking, maybe adding delegates also to replace the `**kwargs.`  Probably the `_batch_tfms = ('after_item','before_batch','after_batch' )` should also be explicit for train and valid. If we do so, this method would end up with like 15 arguments. Don't know what the right solution is, but the actual one is confusing and not expressive enough.

**Comment by arampacha** [comment]
Using delegates here is tricky. We want to delegate to `dl_type` which is passed as argument itself. Delegating to `TfmdDL` may be confusing if one uses other dataloader type.
The version proposed here allows to pass transforms as:
```
dls = DataLoaders.from_dsets(train_ds, valid_ds, 
                             after_item=[TrainOnlyTfm, ToTensor],
                             val_after_item=[ToTensor], 
                             after_batch=[IntToFloatTensor, Normalize.from_stats(*imagenet_stats)])
```
Keyword args starting with `val_` will be used for valid dataloader and if no `val_` argument provided, regular argument will be used for both train and valid dl. Knowing this one can finely control the behavior of `from_dsets`. This pattern is used in other places too, so it shouldn't be completely unfamiliar. But I agree this might be not ideal and possibly better solutions can be found.
Would you share examples where this is not expressive enough? We can try to improve solution based on that.

Also I remember Jeremy mentioned it's possible that v3 can possibly switch to using native pytorch `DataLoader`. If that's correct it will impact methods like `.from_sets`.
